File: k9mail/src/main/java/com/fsck/k9/activity/AccountList.java
package com.fsck.k9.activity;

import java.util.ArrayList;
import java.util.List;

import android.os.AsyncTask;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;

import com.fsck.k9.Account;
import com.fsck.k9.BaseAccount;
import com.fsck.k9.FontSizes;
import com.fsck.k9.K9;
import com.fsck.k9.Preferences;
import com.fsck.k9.R;
import com.fsck.k9.search.SearchAccount;


/**
 * Activity displaying the list of accounts.
 *
 * <p>
 * Classes extending this abstract class have to provide an {@link #onAccountSelected(BaseAccount)}
 * method to perform an action when an account is selected.
 * </p>
 */
public abstract class AccountList extends K9ListActivity implements OnItemClickListener {
    private FontSizes mFontSizes = K9.getFontSizes();

    @Override
    public void onCreate(Bundle icicle) {
        super.onCreate(icicle);

        setResult(RESULT_CANCELED);

        setContentView(R.layout.account_list);

        ListView listView = getListView();
        listView.setOnItemClickListener(this);
        listView.setItemsCanFocus(false);
    }

    /**
     * Reload list of accounts when this activity is resumed.
     */
    @Override
    public void onResume() {
        super.onResume();
        new LoadAccounts().execute();
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        BaseAccount account = (BaseAccount) parent.getItemAtPosition(position);
        onAccountSelected(account);
    }

    /**
     * Create a new {@link AccountsAdapter} instance and assign it to the {@link ListView}.
     *
     * @param realAccounts
     *         An array of accounts to display.
     */
    public void populateListView(List<Account> realAccounts) {
        List<BaseAccount> accounts = new ArrayList<BaseAccount>();

        if (displaySpecialAccounts() && !K9.isHideSpecialAccounts()) {
            BaseAccount unifiedInboxAccount = SearchAccount.createUnifiedInboxAccount(this);
            BaseAccount allMessagesAccount = SearchAccount.createAllMessagesAccount(this);

            accounts.add(unifiedInboxAccount);
            accounts.add(allMessagesAccount);
        }

        accounts.addAll(realAccounts);
        AccountsAdapter adapter = new AccountsAdapter(accounts);
        ListView listView = getListView();
        listView.setAdapter(adapter);
        listView.invalidate();
    }

    /**
     * Implementing decide whether or not to display special accounts in the list.
     *
     * @return {@code true}, if special accounts should be listed. {@code false}, otherwise.
     */
    protected abstract boolean displaySpecialAccounts();

    /**
     * This method will be called when an account was selected.
     *
     * @param account
     *         The account the user selected.
     */
    protected abstract void onAccountSelected(BaseAccount account);

    class AccountsAdapter extends ArrayAdapter<BaseAccount> {
        public AccountsAdapter(List<BaseAccount> accounts) {
            super(AccountList.this, 0, accounts);
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            final BaseAccount account = getItem(position);

            final View view;
            if (convertView != null) {
                view = convertView;
            } else {
                view = getLayoutInflater().inflate(R.layout.accounts_item, parent, false);
                view.findViewById(R.id.active_icons).setVisibility(View.GONE);
            }

            AccountViewHolder holder = (AccountViewHolder) view.getTag();
            if (holder == null) {
                holder = new AccountViewHolder();
                holder.description = (TextView) view.findViewById(R.id.description);
                holder.email = (TextView) view.findViewById(R.id.email);
                holder.chip = view.findViewById(R.id.chip);

                view.setTag(holder);
            }

            String description = account.getDescription();
            if (account.getEmail().equals(description)) {
                holder.email.setVisibility(View.GONE);
            } else {
                holder.email.setVisibility(View.VISIBLE);
                holder.email.setText(account.getEmail());
            }

            if (description == null || description.isEmpty()) {
                description = account.getEmail();
            }

            holder.description.setText(description);

            if (account instanceof Account) {
                Account realAccount = (Account) account;
                holder.chip.setBackgroundColor(realAccount.getChipColor());
            } else {
                holder.chip.setBackgroundColor(0xff999999);
            }

            holder.chip.getBackground().setAlpha(255);

            mFontSizes.setViewTextSize(holder.description, mFontSizes.getAccountName());
            mFontSizes.setViewTextSize(holder.email, mFontSizes.getAccountDescription());


            return view;
        }

        class AccountViewHolder {
            public TextView description;
            public TextView email;
            public View chip;
        }
    }

    /**
     * Load accounts in a background thread
     */
    class LoadAccounts extends AsyncTask<Void, Void, List<Account>> {
        @Override
        protected List<Account> doInBackground(Void... params) {
            List<Account> accounts = Preferences.getPreferences(getApplicationContext()).getAccounts();
            return accounts;
        }

        @Override
        protected void onPostExecute(List<Account> accounts) {
            populateListView(accounts);
        }
    }
}

File: app/ui/legacy/src/main/java/com/fsck/k9/activity/AccountList.kt
package com.fsck.k9.activity

import android.os.AsyncTask
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import com.fsck.k9.Account
import com.fsck.k9.BaseAccount
import com.fsck.k9.K9.isShowUnifiedInbox
import com.fsck.k9.Preferences.Companion.getPreferences
import com.fsck.k9.search.SearchAccount.Companion.createUnifiedInboxAccount
import com.fsck.k9.ui.R
import com.google.android.material.textview.MaterialTextView

/**
 * Activity displaying the list of accounts.
 *
 *
 *
 * Classes extending this abstract class have to provide an [.onAccountSelected]
 * method to perform an action when an account is selected.
 *
 */
abstract class AccountList : K9ListActivity(), OnItemClickListener {
    public override fun onCreate(icicle: Bundle?) {
        super.onCreate(icicle)

        setResult(RESULT_CANCELED)

        setLayout(R.layout.account_list)

        val listView = listView
        listView.onItemClickListener = this
        listView.itemsCanFocus = false
    }

    /**
     * Reload list of accounts when this activity is resumed.
     */
    public override fun onResume() {
        super.onResume()
        LoadAccounts().execute()
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        val account = parent.getItemAtPosition(position) as BaseAccount
        onAccountSelected(account)
    }

    /**
     * Create a new [AccountsAdapter] instance and assign it to the [ListView].
     *
     * @param realAccounts
     * An array of accounts to display.
     */
    fun populateListView(realAccounts: List<Account>?) {
        val accounts: MutableList<BaseAccount> = ArrayList()

        if (isShowUnifiedInbox) {
            val unifiedInboxAccount: BaseAccount = createUnifiedInboxAccount()
            accounts.add(unifiedInboxAccount)
        }

        accounts.addAll(realAccounts!!)
        val adapter: AccountsAdapter = AccountsAdapter(accounts)
        val listView = listView
        listView.adapter = adapter
        listView.invalidate()
    }

    /**
     * This method will be called when an account was selected.
     *
     * @param account
     * The account the user selected.
     */
    protected abstract fun onAccountSelected(account: BaseAccount?)

    internal inner class AccountsAdapter(accounts: List<BaseAccount?>?) : ArrayAdapter<BaseAccount?>(
        this@AccountList, 0, accounts!!,
    ) {
        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val account = getItem(position)
            val view = convertView ?: layoutInflater.inflate(R.layout.accounts_item, parent, false)

            var holder: AccountViewHolder? = view.tag as AccountViewHolder
            if (holder == null) {
                holder = AccountViewHolder()
                holder!!.description = view.findViewById<MaterialTextView>(R.id.description)
                holder.email = view.findViewById<MaterialTextView>(R.id.email)
                holder.chip = view.findViewById<View>(R.id.chip)

                view.tag = holder
            }

            val accountName = account!!.name
            if (accountName != null) {
                holder!!.description!!.text = accountName
                holder.email!!.text = account.email
                holder.email!!.visibility = View.VISIBLE
            } else {
                holder!!.description!!.text = account.email
                holder.email!!.visibility = View.GONE
            }

            if (account is Account) {
                holder.chip!!.setBackgroundColor(account.chipColor)
            } else {
                holder.chip!!.setBackgroundColor(0xff999999.toInt())
            }

            holder.chip!!.background.alpha = 255


            return view
        }

        internal inner class AccountViewHolder {
            var description: MaterialTextView? = null
            var email: MaterialTextView? = null
            var chip: View? = null
        }
    }

    /**
     * Load accounts in a background thread
     */
    internal inner class LoadAccounts : AsyncTask<Void?, Void?, List<Account>>() {
        override fun doInBackground(vararg params: Void?): List<Account> {
            return getPreferences().getAccounts()
        }

        override fun onPostExecute(accounts: List<Account>) {
            populateListView(accounts)
        }
    }
}
